sw/host/opentitanlib/src/bootstrap/mod.rs

```diff
@@ -33,42 +40,81 @@ arg_enum! {
         Eeprom,
+        Rescue,
         Emulator,
     }
 }
 
 // Implementations of bootstrap need to implement the `UpdateProtocol` trait.
 trait UpdateProtocol {
-    fn update(&self, spi: &dyn Target, payload: &[u8]) -> Result<()>;
+    /// Called before any action is taken, to allow the protocol to verify that the transport
+    /// supports SPI/UART or whatever it needs.
+    fn verify_capabilities(
+        &self,
+        container: &Bootstrap,
+        transport: &TransportWrapper,
+    ) -> Result<()>;
+    /// Indicates whether the caller should assert the bootstrap pin and reset the chip, before
+    /// invoking update().
+    fn uses_common_bootstrap_reset(&self) -> bool;
+    /// Invoked to perform the actual transfer of an executable image to the OpenTitan chip.
+    fn update(
+        &self,
+        container: &Bootstrap,
+        transport: &TransportWrapper,
+        payload: &[u8],
+    ) -> Result<()>;
 }
 
 /// Options which control bootstrap behavior.
 /// The meaning of each of these values depends on the specific bootstrap protocol being used.
-#[derive(Debug, Default)]
+#[derive(Debug, StructOpt)]
 pub struct BootstrapOptions {
-    /// How long to hold the reset pin during the reset sequence.
+    #[structopt(flatten)]
+    pub uart_params: UartParams,
+    #[structopt(flatten)]
+    pub spi_params: SpiParams,
+    #[structopt(
+        short,
+        long,
+        possible_values = &BootstrapProtocol::variants(),
+        case_insensitive = true,
+        default_value = "primitive",
+        help = "Bootstrap protocol to use"
+    )]
+    pub protocol: BootstrapProtocol,
+    #[structopt(long, parse(try_from_str=parse_duration), help = "Duration of the reset delay")]
     pub reset_delay: Option<Duration>,
-    /// How long to delay between sending bootstrap frames.
+    #[structopt(long, parse(try_from_str=parse_duration), help = "Duration of the inter-frame delay")]
     pub inter_frame_delay: Option<Duration>,
-    /// How long to delay during a flash erase operation.
+    #[structopt(long, parse(try_from_str=parse_duration), help = "Duration of the flash-erase delay")]
     pub flash_erase_delay: Option<Duration>,
 }
 
 /// Bootstrap wraps and drives the various bootstrap protocols.
-pub struct Bootstrap {
+pub struct Bootstrap<'a> {
     pub protocol: BootstrapProtocol,
+    pub uart_params: &'a UartParams,
+    pub spi_params: &'a SpiParams,
+    reset_pin: Rc<dyn GpioPin>,
+    bootstrap_pin: Rc<dyn GpioPin>,
     reset_delay: Duration,
-    updater: Box<dyn UpdateProtocol>,
 }
 
-impl Bootstrap {
+impl<'a> Bootstrap<'a> {
     const RESET_DELAY: Duration = Duration::from_millis(200);
 
-    /// Consrtuct a `Bootstrap` struct configured to use `protocol` and `options`.
-    pub fn new(protocol: BootstrapProtocol, options: BootstrapOptions) -> Result<Self> {
-        let updater: Box<dyn UpdateProtocol> = match protocol {
+    /// Perform the update, sending the firmware `payload` to the `spi` target,
```

Comment:
nit: Can this a UART target in the case of the new "rescue" protocol?

Reply:
Oh yes, I forgot to update this comment, leaving it referring to parameters that no longer exist.  I have updated it.